src/fstab-generator/fstab-generator.c

```diff
@@ -554,27 +554,63 @@ static int parse_fstab(bool initrd) {
+                        _cleanup_free_ char
+                                *absolute_what = NULL,
+                                *canonical_what = NULL;
+
+                        path_kill_slashes(what);
+
+                        /* Both mount and fstab-generator have accepted relative paths.
+                         * Stay compatible and allow the system to boot. */
+                        absolute_what = path_make_absolute(what, "/");
+                        if (!absolute_what)
+                                return log_oom();
 
-                if (is_path(where)) {
-                        path_kill_slashes(where);
-                        /* Follow symlinks here; see 5261ba901845c084de5a8fd06500ed09bfb0bd80 which makes sense for
-                         * mount units, but causes problems since it historically worked to have symlinks in e.g.
-                         * /etc/fstab. So we canonicalize here. Note that we use CHASE_NONEXISTENT to handle the case
-                         * where a symlink refers to another mount target; this works assuming the sub-mountpoint
-                         * target is the final directory.
-                         */
-                        r = chase_symlinks(where, initrd ? "/sysroot" : NULL,
+                        /* Similarly, canonicalize the source directory of bind mounts.
+                         * In particular this makes a best-effort attempt to keep the
+                         * source outside of the initrd, e.g. if "/.." is specified. */
+                        r = chase_symlinks(absolute_what, initrd ? "/sysroot" : NULL,
                                            CHASE_PREFIX_ROOT | CHASE_NONEXISTENT,
-                                           &canonical_where);
+                                           &canonical_what);
                         if (r < 0)
                                 /* In this case for now we continue on as if it wasn't a symlink */
-                                log_warning_errno(r, "Failed to read symlink target for %s: %m", where);
+                                log_warning_errno(r, "Failed to read symlink target for %s: %m", what);
                         else {
-                                if (streq(canonical_where, where))
-                                        canonical_where = mfree(canonical_where);
-                                else
-                                        log_debug("Canonicalized what=%s where=%s to %s",
-                                                  what, where, canonical_where);
+                                if (!streq(canonical_what, what)) {
+                                        log_debug("Canonicalized what=%s to %s (for where=%s)",
+                                                  what, canonical_what, where);
+                                        free_and_replace(what, canonical_what);
+                                        canonical_what = NULL;
```

Comment:
`free_and_replace()` already sets `canonical_what = NULL`.

Reply:
Even better, thank you.